students/smitco/lesson05/mailroom3.py

```diff
@@ -0,0 +1,123 @@
+# lesson 04 mailroom pt 2
+# !/usr/bin/env python3
+
+
+import os
+import datetime
+
+
+# starting list of 5+ donors and 1-3 donations each
+donors = {"John Travolta": [5000, 7500],
+          "Jane Fonda": [10000, 8000, 6500],
+          "Judy Blume": [3000, 3000],
+          "Joey Tribbiani": [9000],
+          "Jenny Gump": [10300, 13750, 12500]}
+
+
+def menu():
+    ask = input("Please choose an action:\n"
+                "1) Send a new thank you\n"
+                "2) Create a report\n"
+                "3) Send letters to all donors\n"
+                "4) Quit\n"
+                ">>")
+    options = {"1": donor_name, "2": report, "3": letters, "4": quit}
+    try:  
+        options[ask]()
+    except KeyError:
+        print("\nThere was an error. Please try again.\n")
+
+
+def email(name, donation):
+    """Print short email to screen"""
+    if donation >= (10 ** 6):
+        print("\nThe amount entered is too large.")
+        donors.pop(name)
+    else:
+        donors[name].append(donation)
+        ty_temp = {"first_last": name, "value": donation}
+        print("\nThank you, {first_last}, for supporting The Brave Heart Foundation.\n"
+              "Your generous donation of ${value} will make a positive,\n"
+              "life-changing impact for teens nationwide.".format(**ty_temp))
+
+
+def donation_amount(donor):
+    """Ask for donor amount"""
+    donation = input("\nWhat is the donation amount?\n>>")
+    if donation.lower() == "exit":
+        print("\nExiting.")
+        donors.pop(donor)
+    else:
+        try:
+            donation = int(donation)
+        except ValueError:
+            print("\nInvalid entry.")
+            donors.pop(donor)
+        else:
+            email(donor, donation)
+
+
+def donor_name():
+    """Ask for donor name and add to list if needed"""
+    while True:
+        addressee = input("\nTo whom would you like to send a thank you?\n"
+                          "'List' will display current donors.\n"
+                          "'Exit' will return to main menu.\n"
+                          ">>")
+        if addressee.lower() == "exit":
+            print("\nExiting.\n")
+            break
+        elif addressee.lower() == "list":
+            print(list(donors.keys()))
+        elif addressee not in donors:
+            donors[addressee] = []
+            donation_amount(addressee)
+        else:
+            donation_amount(addressee)
+
+
+def report():
+    """Print report of donors and donation details"""
+    donor_sums = {}
+    for d in donors:
+        total = sum(donors[d])
+        count = len(donors[d])
+        average = total/count
+        donor_sums[d] = [total, count, average]
```

Comment:
out of the lines you mentioned in your comment this is the only one that could be a good candidate for comprehensions; you want to avoid losing a lot of readability just for the sake of one-liners 🙂 

```
donor_sums = {k: [sum(v), len(v), sum(v) / len(v)] for d in donors.items()}
```